File: cli/lib/dobase/commands/account.rb
# frozen_string_literal: true

require "io/console"

module Dobase
  module Commands
    class Account < Command
      command "login", "Save the server URL and an access token for this machine", args: %w[[URL]] do |url = nil|
        url = (url || config.url || ask("Dobase URL: ")).to_s.strip.chomp("/")
        url = "https://#{url}" unless url.match?(%r{\Ahttps?://})

        $stderr.puts "Create an access token under Profile → API: #{url}/profile/edit?tab=api"
        token = ask("Access token: ", secret: true).to_s.strip
        raise UsageError, "No token given." if token.empty?

        @client = Client.new(url: url, token: token, user_agent: @user_agent)
        profile = get("/profile")
        config.save(url: url, token: token)

        say "Signed in to #{url} as #{person(profile)}."
        say "Token #{quoted(profile.dig("access_token", "name"))} can #{profile.dig("access_token", "permission") == "write" ? "read and write" : "only read"}."
      end

      command "logout", "Forget the saved URL and token (revoke the token under Profile → API)" do
        config.forget
        say "Signed out. The token still works until you revoke it under Profile → API."
      end

      command "whoami", "Show who the token belongs to and what it may do" do
        output(me) do |profile|
          token = profile["access_token"]
          say "#{person(profile)} on #{config.url}"
          say "Token #{quoted(token["name"])} (#{token["permission"] == "write" ? "read and write" : "read only"})" if token
        end
      end

      private

      # Prompts only when someone is typing; a piped token is read silently.
      def ask(prompt, secret: false)
        return $stdin.gets unless $stdin.tty?

        $stderr.print prompt
        value = secret ? $stdin.noecho(&:gets) : $stdin.gets
        $stderr.puts if secret
        value
      end
    end
  end
end

File: cli/src/commands/account.rs
use std::io::{BufRead, IsTerminal, Write};

use crate::client::Client;
use crate::command::{Args, Ctx, Definition, Error, Result, command, person, quoted, usage};
use crate::value::Json;

pub fn definitions() -> Vec<Definition> {
    vec![
        command("login", "Save the server URL and an access token for this machine", &["[URL]"], vec![], login),
        command("logout", "Forget the saved URL and token (revoke the token under Profile → API)", &[], vec![], logout),
        command("whoami", "Show who the token belongs to and what it may do", &[], vec![], whoami),
    ]
}

fn login(ctx: &mut Ctx, args: &Args) -> Result<()> {
    let url = match args.get(0).map(str::to_string).or_else(|| ctx.config.url()) {
        Some(url) => url,
        None => ask("Dobase URL: ", false)?,
    };
    let mut url = url.trim().trim_end_matches('/').to_string();
    if !url.starts_with("http://") && !url.starts_with("https://") {
        url = format!("https://{url}");
    }

    eprintln!("Create an access token under Profile → API: {url}/profile/edit?tab=api");
    let token = ask("Access token: ", true)?.trim().to_string();
    if token.is_empty() {
        usage!("No token given.");
    }

    let client = Client::new(Some(url.clone()), Some(token.clone()), &ctx.user_agent)?;
    ctx.set_api(Box::new(client));
    let profile = ctx.get("/profile", &[])?;
    ctx.config.save(&url, &token)?;

    ctx.say(format!("Signed in to {url} as {}.", person(&profile).unwrap_or_default()));
    let permission = if profile["access_token"]["permission"].s() == "write" { "read and write" } else { "only read" };
    ctx.say(format!("Token {} can {permission}.", quoted(&profile["access_token"]["name"].s())));
    Ok(())
}

fn logout(ctx: &mut Ctx, _args: &Args) -> Result<()> {
    ctx.config.forget()?;
    ctx.say("Signed out. The token still works until you revoke it under Profile → API.");
    Ok(())
}

fn whoami(ctx: &mut Ctx, _args: &Args) -> Result<()> {
    let profile = ctx.me()?;
    ctx.output(&profile, |ctx| {
        let url = ctx.config.url().unwrap_or_default();
        ctx.say(format!("{} on {url}", person(&profile).unwrap_or_default()));
        let token = &profile["access_token"];
        if !token.is_null() {
            let permission = if token["permission"].s() == "write" { "read and write" } else { "read only" };
            ctx.say(format!("Token {} ({permission})", quoted(&token["name"].s())));
        }
        Ok(())
    })
}

/// Prompts only when someone is typing; a piped token is read silently.
fn ask(prompt: &str, secret: bool) -> Result<String> {
    let stdin = std::io::stdin();
    let read_line = || {
        let mut line = String::new();
        stdin.lock().read_line(&mut line).map(|_| line)
    };

    let answer = if !stdin.is_terminal() {
        read_line()
    } else {
        eprint!("{prompt}");
        let _ = std::io::stderr().flush();
        if secret { rpassword::read_password() } else { read_line() }
    };
    answer.map_err(|error| Error::failed(format!("Could not read the answer: {error}")))
}
